File: linkedin-scraper/browser/python/linkedin.py
"""LinkedIn scraper using the official Scrapeless Python SDK + Playwright over CDP.
**Public surfaces only.** LinkedIn aggressively walls off logged-in content;
this scraper only touches:
  - public profile pages              -> https://linkedin.com/in/<slug>
  - public company pages              -> https://linkedin.com/company/<id> (and /life)
  - the unauthenticated jobs guest API -> /jobs-guest/jobs/api/seeMoreJobPostings/search
  - job posting pages                  -> /jobs/view/<id>
  - articles                           -> /pulse/<slug>
"""

from __future__ import annotations

import json
import os
from typing import Any
from urllib.parse import quote

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _fetch_rendered_html(
    url: str,
    ready_selector: str | None = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 1,
) -> str:
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        ws = session.browser_ws_endpoint
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(ws)
            try:
                page = await browser.new_page()
                await page.goto(url, wait_until="domcontentloaded", timeout=30000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=15000)
                    except Exception as e:
                        logger.warning("wait_for_selector failed (continuing): {}", e)
                html = await page.content()
                if html:
                    return html
                last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

def _strip_text(value: str | None) -> str:
    return (value or "").strip()

def _find_ld_node(graph: list[dict[str, Any]], type_name: str) -> dict[str, Any] | None:
    for node in graph:
        t = node.get("@type")
        if t == type_name or (isinstance(t, list) and type_name in t):
            return node
    return None

# ---------------------------------------------------------------------------
# Parsers — mirror the upstream reference's keys verbatim
# ---------------------------------------------------------------------------

def parse_profile(html: str) -> dict[str, Any]:
    sel = Selector(text=html)
    blocks = sel.xpath("//script[@type='application/ld+json']/text()").getall()
    graph: list[dict[str, Any]] = []
    for raw in blocks:
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if isinstance(data, dict) and "@graph" in data:
            graph.extend(data["@graph"])
        elif isinstance(data, list):
            graph.extend(data)
        elif isinstance(data, dict):
            graph.append(data)
    profile = _find_ld_node(graph, "Person") or {}
    posts = [n for n in graph if (
        n.get("@type") == "Article"
        or (isinstance(n.get("@type"), list) and "Article" in n["@type"])
    )]
    return {"profile": profile, "posts": posts}

def parse_company(html: str) -> dict[str, Any]:
    sel = Selector(text=html)
    blocks = sel.xpath("//script[@type='application/ld+json']/text()").getall()
    org: dict[str, Any] = {}
    for raw in blocks:
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if isinstance(data, dict):
            t = data.get("@type")
            if t == "Organization" or (isinstance(t, list) and "Organization" in t):
                org = data
                break
            if "@graph" in data:
                cand = _find_ld_node(data["@graph"], "Organization")
                if cand:
                    org = cand
                    break

    out: dict[str, Any] = {
        "name": org.get("name", ""),
        "url": org.get("url", ""),
        "mainAddress": org.get("address"),
        "description": org.get("description"),
        "numberOfEmployees": org.get("numberOfEmployees"),
        "logo": org.get("logo"),
    }
    # Lift the about-us key/value pairs (Industry, Company size, etc.).
    for row in sel.xpath("//div[contains(@data-test-id, 'about-us')]//dl/div"):
        key = _strip_text(row.xpath(".//dt//text()").get())
        val = _strip_text(" ".join(row.xpath(".//dd//text()").getall()))
        if key:
            out[key] = val

    # /life surface extras — only present if URL ends in /life.
    leaders = []
    for li in sel.xpath("//section[@data-test-id='leaders-at']//li"):
        leaders.append({
            "name": _strip_text(li.xpath(".//h3//text()").get()),
            "linkedinProfileLink": li.xpath(".//a/@href").get() or "",
        })
    if leaders:
        out["leaders"] = leaders

    def _page_list(xp: str) -> list[dict[str, Any]]:
        items = []
        for li in sel.xpath(xp):
            items.append({
                "name": _strip_text(li.xpath(".//h3//text()").get()),
                "industry": _strip_text(li.xpath(".//p[1]//text()").get()),
                "address": _strip_text(li.xpath(".//p[2]//text()").get()),
                # the upstream reference emits this key with a typo — preserve it.
                "linkeinUrl": li.xpath(".//a/@href").get() or "",
            })
        return items

    aff = _page_list("//section[@data-test-id='affiliated-pages']/div/div/ul/li")
    sim = _page_list("//section[@data-test-id='similar-pages']/div/div/ul/li")
    if aff:
        out["affiliatedPages"] = aff
    if sim:
        out["similarPages"] = sim
    return out

def parse_job_search(html: str) -> dict[str, Any]:
    sel = Selector(text=html)
    # Current public layout: ul.jobs-search__results-list > li > div.base-search-card
    cards = sel.xpath(
        "//ul[contains(@class, 'jobs-search__results-list')]/li[.//div[contains(@class, 'base-search-card')] or .//h3]"
    )
    if not cards:
        cards = sel.xpath("//section[contains(@class, 'results-list')]/ul/li")
    if not cards:
        cards = sel.xpath("//li[contains(@class, 'jobs-search-results__list-item') or contains(@class, 'result-card')]")
    if not cards:
        cards = sel.xpath("//div[contains(@class, 'base-search-card') or contains(@class, 'job-search-card')]")
    data = []
    for li in cards:
        title = _strip_text(li.xpath(".//h3//text()").get())
        company = _strip_text(li.xpath(".//h4//text()").get())
        if not title and not company:
            continue
        data.append({
            "title": title,
            "company": company,
            "address": _strip_text(li.xpath(".//*[contains(@class, 'job-search-card__location')]//text()").get()),
            "timeAdded": li.xpath(".//time/@datetime").get() or "",
            "jobUrl": li.xpath(".//a[contains(@class, 'base-card__full-link')]/@href").get()
                      or li.xpath(".//a/@href").get() or "",
            "companyUrl": li.xpath(".//h4//a/@href").get() or "",
        })
    total_text = sel.xpath("//span[contains(@class, 'results-context-header__job-count')]//text()").get()
    total = None
    if total_text:
        try:
            total = int("".join(c for c in total_text if c.isdigit()))
        except ValueError:
            total = None
    return {"data": data, "total_results": total}

def parse_job_page(html: str) -> dict[str, Any]:
    sel = Selector(text=html)
    blocks = sel.xpath("//script[@type='application/ld+json']/text()").getall()
    payload: dict[str, Any] = {}
    for raw in blocks:
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if isinstance(data, dict):
            t = data.get("@type")
            if t == "JobPosting" or (isinstance(t, list) and "JobPosting" in t):
                payload = data
                break
    bullets = sel.xpath("//div[contains(@class, 'show-more')]/ul/li/text()").getall()
    if bullets:
        payload["jobDescription"] = [_strip_text(b) for b in bullets if _strip_text(b)]
    return payload

def parse_article_page(html: str) -> dict[str, Any]:
    sel = Selector(text=html)
    blocks = sel.xpath("//script[@type='application/ld+json']/text()").getall()
    payload: dict[str, Any] = {}
    for raw in blocks:
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        if isinstance(data, dict):
            t = data.get("@type")
            if t == "Article" or (isinstance(t, list) and "Article" in t):
                payload = data
                break
    spans = sel.xpath(
        "//article/div[@data-test-id='article-content-blocks']/div/p/span/text()"
    ).getall()
    if spans:
        payload["articleBody"] = "\n".join(_strip_text(s) for s in spans if _strip_text(s))
    return payload

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference's exports verbatim
# ---------------------------------------------------------------------------

async def scrape_profile(urls: list[str]) -> list[dict[str, Any]]:
    """Scrape public LinkedIn profile pages."""
    out = []
    for url in urls:
        html = await _fetch_rendered_html(url, ready_selector="script[type='application/ld+json']")
        out.append(parse_profile(html))
    return out

async def scrape_company(urls: list[str]) -> list[dict[str, Any]]:
    """Scrape public LinkedIn company pages."""
    out = []
    for url in urls:
        html = await _fetch_rendered_html(url, ready_selector="script[type='application/ld+json']")
        out.append(parse_company(html))
    return out

async def scrape_job_search(
    keyword: str, location: str, max_pages: int | None = None
) -> list[dict[str, Any]]:
    """Scrape the LinkedIn jobs guest API."""
    first_url = (
        f"https://www.linkedin.com/jobs/search?keywords={quote(keyword)}"
        f"&location={quote(location)}"
    )
    first_html = await _fetch_rendered_html(first_url, ready_selector="section.results-list, ul")
    first_page = parse_job_search(first_html)
    pages: list[dict[str, Any]] = [first_page]
    pages_to_fetch = max_pages if max_pages else 1
    # Guest pagination uses /jobs-guest/jobs/api/seeMoreJobPostings/search?...&start=N
    for i in range(1, pages_to_fetch):
        start = i * 25
        url = (
            "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search"
            f"?keywords={quote(keyword)}&location={quote(location)}&start={start}"
        )
        try:
            html = await _fetch_rendered_html(url)
            pages.append(parse_job_search(html))
        except Exception as e:  # noqa: BLE001
            logger.warning("job-search pagination failed at start={}: {}", start, e)
            break
    return pages

async def scrape_jobs(urls: list[str]) -> list[dict[str, Any]]:
    """Scrape LinkedIn job pages."""
    out = []
    for url in urls:
        html = await _fetch_rendered_html(url, ready_selector="script[type='application/ld+json']")
        out.append(parse_job_page(html))
    return out

async def scrape_articles(urls: list[str]) -> list[dict[str, Any]]:
    """Scrape LinkedIn articles (`/pulse/...`)."""
    out = []
    for url in urls:
        html = await _fetch_rendered_html(url, ready_selector="script[type='application/ld+json']")
        out.append(parse_article_page(html))
    return out

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: linkedin-scraper/browser/nodejs/linkedin.mjs
// LinkedIn scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// the function names and emitted field names match verbatim.
//
// Public surfaces only: profiles, company pages (incl. /life), the unauthenticated
// jobs guest API, job pages, and Pulse articles. No authenticated scraping.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}
function client() { return new Scrapeless({ apiKey: requireKey() }); }

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 1 } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 15000 }); } catch (_) {}
      }
      const html = await page.content();
      if (html) return html;
      lastError = new Error("empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

function readJsonLdGraph($) {
  const graph = [];
  $('script[type="application/ld+json"]').each((_, el) => {
    const raw = $(el).contents().text();
    try {
      const data = JSON.parse(raw);
      if (data && typeof data === "object") {
        if (Array.isArray(data)) graph.push(...data);
        else if (Array.isArray(data["@graph"])) graph.push(...data["@graph"]);
        else graph.push(data);
      }
    } catch (_) {}
  });
  return graph;
}
function findNode(graph, typeName) {
  return graph.find((n) => {
    const t = n["@type"];
    return t === typeName || (Array.isArray(t) && t.includes(typeName));
  });
}

// ---------------- parsers ----------------

export function parseProfile(html) {
  const $ = cheerio.load(html);
  const graph = readJsonLdGraph($);
  return {
    profile: findNode(graph, "Person") ?? {},
    posts: graph.filter((n) => {
      const t = n["@type"];
      return t === "Article" || (Array.isArray(t) && t.includes("Article"));
    }),
  };
}

export function parseCompany(html) {
  const $ = cheerio.load(html);
  const graph = readJsonLdGraph($);
  const org = findNode(graph, "Organization") ?? {};
  const out = {
    name: org.name ?? "",
    url: org.url ?? "",
    mainAddress: org.address ?? null,
    description: org.description ?? null,
    numberOfEmployees: org.numberOfEmployees ?? null,
    logo: org.logo ?? null,
  };
  $('[data-test-id*="about-us"] dl > div').each((_, row) => {
    const key = $(row).find("dt").text().trim();
    const val = $(row).find("dd").text().trim();
    if (key) out[key] = val;
  });

  const leaders = [];
  $('[data-test-id="leaders-at"] li').each((_, li) => {
    leaders.push({
      name: $(li).find("h3").first().text().trim(),
      linkedinProfileLink: $(li).find("a").first().attr("href") ?? "",
    });
  });
  if (leaders.length) out.leaders = leaders;

  const collect = (sel) => {
    const items = [];
    $(sel).each((_, li) => {
      items.push({
        name: $(li).find("h3").first().text().trim(),
        industry: $(li).find("p").eq(0).text().trim(),
        address: $(li).find("p").eq(1).text().trim(),
        linkeinUrl: $(li).find("a").first().attr("href") ?? "",
      });
    });
    return items;
  };
  const aff = collect('[data-test-id="affiliated-pages"] ul > li');
  const sim = collect('[data-test-id="similar-pages"] ul > li');
  if (aff.length) out.affiliatedPages = aff;
  if (sim.length) out.similarPages = sim;
  return out;
}

export function parseJobSearch(html) {
  const $ = cheerio.load(html);
  // LinkedIn's public jobs search renders one <li> per card under
  // ul.jobs-search__results-list with a div.base-search-card inside it. Older
  // layouts used section.results-list ul > li.
  let cards = $("ul.jobs-search__results-list > li").filter((_, li) => $(li).find("div.base-search-card, a.base-card__full-link, h3").length > 0);
  if (!cards.length) cards = $("section.results-list ul > li");
  if (!cards.length) cards = $("li.jobs-search-results__list-item, li.result-card");
  if (!cards.length) cards = $("div.base-search-card, div.job-search-card");
  const data = [];
  cards.each((_, li) => {
    const $li = $(li);
    const title = $li.find("h3.base-search-card__title, h3").first().text().trim();
    const company = $li.find("h4.base-search-card__subtitle a, h4.base-search-card__subtitle, h4").first().text().trim();
    if (!title && !company) return;
    data.push({
      title,
      company,
      address: $li.find(".job-search-card__location").first().text().trim(),
      timeAdded: $li.find("time").first().attr("datetime") ?? "",
      jobUrl: $li.find("a.base-card__full-link").first().attr("href")
        ?? $li.find("a").first().attr("href") ?? "",
      companyUrl: $li.find("h4 a").first().attr("href") ?? "",
    });
  });
  const totalText = $(".results-context-header__job-count").first().text();
  let total_results = null;
  if (totalText) {
    const digits = totalText.replace(/\D/g, "");
    if (digits) total_results = Number(digits);
  }
  return { data, total_results };
}

export function parseJobPage(html) {
  const $ = cheerio.load(html);
  const graph = readJsonLdGraph($);
  const payload = findNode(graph, "JobPosting") ?? {};
  const bullets = $("div.show-more ul > li")
    .map((_, li) => $(li).text().trim())
    .get()
    .filter(Boolean);
  if (bullets.length) payload.jobDescription = bullets;
  return payload;
}

export function parseArticlePage(html) {
  const $ = cheerio.load(html);
  const graph = readJsonLdGraph($);
  const payload = findNode(graph, "Article") ?? {};
  const parts = $('article div[data-test-id="article-content-blocks"] div p span')
    .map((_, el) => $(el).text().trim())
    .get()
    .filter(Boolean);
  if (parts.length) payload.articleBody = parts.join("\n");
  return payload;
}

// ---------------- scrape functions ----------------

export async function scrapeProfile(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseProfile(html));
  }
  return out;
}

export async function scrapeCompany(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseCompany(html));
  }
  return out;
}

export async function scrapeJobSearch(keyword, location, maxPages = null) {
  const first = `https://www.linkedin.com/jobs/search?keywords=${encodeURIComponent(keyword)}&location=${encodeURIComponent(location)}`;
  const firstHtml = await fetchRenderedHtml(first, "section.results-list, ul");
  const pages = [parseJobSearch(firstHtml)];
  const toFetch = maxPages ?? 1;
  for (let i = 1; i < toFetch; i++) {
    const start = i * 25;
    const url = `https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search?keywords=${encodeURIComponent(keyword)}&location=${encodeURIComponent(location)}&start=${start}`;
    try {
      const html = await fetchRenderedHtml(url);
      pages.push(parseJobSearch(html));
    } catch (_) { break; }
  }
  return pages;
}

export async function scrapeJobs(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseJobPage(html));
  }
  return out;
}

export async function scrapeArticles(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseArticlePage(html));
  }
  return out;
}
